config/sources/common.conf

```diff
@@ -6,19 +6,180 @@
 # This file is a part of the Armbian Build Framework
 # https://github.com/armbian/build/
 #
+
 # this is sourced before any other arch specific config file, always. see main-config.sh
 
+# Source the mainline kernel hooks, in a separate file for easier maintenance.
+
+# shellcheck source=/dev/null
+source "${SRC}/config/sources/mainline-kernel.conf.sh"
+
 declare -g FAST_CREATE_IMAGE='yes'
 declare -g MAIN_CMDLINE='rw no_console_suspend consoleblank=0 fsck.fix=yes fsck.repair=yes net.ifnames=0 splash plymouth.ignore-serial-consoles'
 
-# boot loader configuration
-[[ -z $BOOTSOURCE ]] && declare -g BOOTSOURCE="$MAINLINE_UBOOT_SOURCE"
-[[ -z $BOOTDIR ]] && declare -g BOOTDIR="$MAINLINE_UBOOT_DIR"
-[[ -z $BOOTBRANCH ]] && declare -g BOOTBRANCH="${BOOTBRANCH_BOARD:-"tag:v2022.07"}"
+# So the idea is to let the board/family/hooks define these variables.
+# If the family does NOT define them, assume sane defaults.
+function late_family_config__common_defaults_for_mainline_kernel() {
+	# @TODO: if KERNELSOURCE is 'none', don't do anything; config is explicitly asking to skip kernel compilation.
+	if [[ "${KERNELSOURCE}" == "none" ]]; then
```

Comment:
This whole function looks like a validation function for check if all the required things are present or not. I think this should be moved out of config and to libs perhaps

Reply:
Before this PR, having KERNELSOURCE unset meant `don't build a kernel`.
After this PR, having KERNELSOURCE unset means `use a mainline kernel`, and to avoid building a kernel, KERNELSOURCE must be set to `none`. This is the same logic as was used for BOOTCONFIG=none to avoid building u-boot.
Also, yes, those are hooks (functions) and can be moved around from config to libs or anywhere that is sourced before exts initialize.